### DISCUSSION TITLE
2. Soruda equals kullanımı

### DISCUSSION THREAD

**Discussion by mesutckr** [gusanmaz/BMB212-Algorithms-2223#9] [Q&A] [answered]
Merhaba,
2. soruda string karşılaştırmak için == kullandığımızda sorun yaşıyorum. == yerine equals() kullanmak daha doğru olmaz mı? bu soruda equals() kullansak sorun olur mu?

**Comment by gusanmaz** [comment]
~~`==` operatoru ile `equals` metodu ayni islevi goruyor ama odevlerinizde `==` operatorunu kullanmaniz isteniyor.~~

```java
class Main {
  public static void main(String[] args) {
    String s1 = "fil";
    String s2 = "fil";
    String s3 =  "zil";
    String s4 =  "bil";

    System.out.println("==");
    System.out.println(s1 == s2);
    System.out.println(s1 == s3);
    System.out.println(s4 == s4);
    System.out.println(s1 == s4);
    System.out.println(s2 == s3);

    System.out.println();

    System.out.println("equals");
    System.out.println(s1.equals(s2));
    System.out.println(s1.equals(s3));
    System.out.println(s4.equals(s4));            
    System.out.println(s1.equals(s4));
    System.out.println(s2.equals(s3));
  }
}
```

Ornegin yukaridaki kodun ciktisi asagidaki gibi oluyor:

```bash
==
true
false
true
false
false

equals
true
false
true
false
false
```

Ilgili kodu https://replit.com/@GuvencUsanmaz/StringComparison baglantisindan canli deneyebilirsiniz.

**Comment by mesutckr** [comment]
== operatoru ile equals metodu aynı işlevi görmez. == string değişkenleri referans olarak karşılaştırır equals() ise  string değişkenleri değer olarak karşılaştırır.  sizin örneğinizde sadece s2 değişkeninin yeniden referans oluşturularak yaratılmasını istersek işlerin nasıl değiştiğini kendinizde deneyerek görebilirsiniz. s1 ve s2 değerleri aynıdır fakat referansları farklıdır.

            String s1 = "fil";
	    String s2 = new String("fil");
	    String s3 =  "zil";
	    String s4 =  "bil";
	   

	    System.out.println("==");
	    System.out.println(s1 == s2);
	    System.out.println(s1 == s3);
	    System.out.println(s4 == s4);
	    System.out.println(s1 == s4);
	    System.out.println(s2 == s3);

	    System.out.println();

	    System.out.println("equals");
	    System.out.println(s1.equals(s2));
	    System.out.println(s1.equals(s3));
	    System.out.println(s4.equals(s4));            
	    System.out.println(s1.equals(s4));
	    System.out.println(s2.equals(s3));


ÇIKTILAR
              ==
              false
              false
              true
              false
              false
              
              equals
              true
              false
              true
              false
              false

Çıktılardan da değerleri karşılaştırmak için equals() kullanılması gerektiğini görebiliriz.
bundan yola çıkarak bu soruda equals() kullanmak daha doğru olacaktır...

**Comment by gusanmaz** [comment, marked as answer]
Evet. string karsilastirma icin `==` operatorunu kullanmaniz dogru olmaz. String iceriklerini karsilastirmak icin `equals` metodunu kulanabilirsiniz. `README` dosyasinda gerekli degisiklik yapildi.
